File: 민재/자료구조/스택/4949_균형잡힌세상.swift
//
//  main.swift
//  algo
//
//  Created by 김민재 on 2022/05/23.
//

import Foundation

var input = ""

while true {
    input = readLine()!
    if input == "." { break }
    
    var stack: [Character] = []
    var balanceFlag = true
    
    for char in input {
        if char == "(" || char == "[" {
            stack.append(char)
        } else if char == ")" {
            if !stack.isEmpty && stack[stack.count-1] == "(" {
                stack.removeLast()
            } else {
                // stack이 비어있다
                balanceFlag = false
                break
            }
        } else if char == "]" {
            if !stack.isEmpty && stack[stack.count-1] == "[" {
                stack.removeLast()
            } else {
                balanceFlag = false
                break
            }
        }
    }
    
    if !stack.isEmpty || !balanceFlag {
        print("no")
    } else {
        print("yes")
    }
    
    
}

File: 영진/자료 구조/4949_균형잡힌 세상.py
import sys
input = sys.stdin.readline
while True:
    stack = []

    script = input()
    if script == '.':
        break

    for ch in script:
        if ch == '(' or ch == '[':
            stack.append(ch)
        elif ch == ')':
            if len(stack) != 0 and stack[-1] == '(':
                stack.pop()
            else:
                stack.append(ch)
        elif ch == ']':
            if len(stack) != 0 and stack[-1] == '[':
                stack.pop()
            else:
                stack.append(ch)

    if len(stack) == 0:
        print("yes")
    else:
        print("no")
